abel/direct.py

```diff
@@ -209,28 +205,28 @@ def _pyabel_direct_integral(f, r, correction, int_func=np.trapz):
 
 
     """
-    Compute the correction
-    Pre-calculated analytical integration of the cell with the singular value
-    Assuming a piecewise linear behaviour of the data
+    Compute the correction. Here we apply an
+    analytical integration of the cell with the singular value,
+    assuming a piecewise linear behaviour of the data.
+    The analytical abel transform for this trapezoid is:
     c0*acosh(r1/y) - c_r*y*acosh(r1/y) + c_r*sqrt(r1**2 - y**2)
+    see: https://github.com/luli/hedp/blob/master/hedp/math/abel.py#L87-L104
     """
     if correction == 1:
-        # computing forward derivative of the data
-        f_r = (f[:, 1:] - f[:, :N1-1])/np.diff(r)[None, :]
+
+        # precompute a few variables outside the loop:
+        f_r = (f[:, 1:] - f[:, :-1])/np.diff(r)[None, :]
+        isqrt = I_sqrt[II+1 == JJ]
+
+        if np.isclose(r[0], 0):  # special case for r[0] = 0
+            ratio = np.append(np.cosh(1), r[2:]/r[1:-1])
+        else:
+            ratio = r[1:]/r[:-1]
+        
+        acr = np.arccosh(ratio)
 
         for i, row in enumerate(f):  # loop over rows (z)
-            corr  = I_sqrt[II+1 == JJ]*f_r[i] \
-                    + np.arccosh(r[1:]/r[:-1])*(row[:-1] - f_r[i]*r[:-1])
-
-            
-            if np.isclose(r[0], 0):  # special case for r[0] = 0 
-                corr0 = I_sqrt[II+1 == JJ][0]*f_r[i,0] \
-                        + (row[0] - f_r[i,0]*r[0])
-                out[i, 1:-1] += corr[1:]
-                out[i, 0]    += corr0 
-            else:
-                out[i, :-1] = corr
-            
+            out[i, :-1] += isqrt*f_r[i] + acr*(row[:-1] - f_r[i]*r[:-1])
 
     return out
 
```

Comment:
The assignment, line 182, `N1 = f.shape[1]` may be deleted, since we now use `[:, -1]` instead of `[:, N1-1]`